packages/iTwinUI-react/src/core/Table/actionHandlers/selectHandler.ts

```diff
@@ -56,6 +56,91 @@ export const onSelectHandler = <T extends Record<string, unknown>>(
   onSelect?.(selectedData, newState);
 };
 
+export const onShiftSelectHandler = <T extends Record<string, unknown>>(
+  state: TableState<T>,
+  action: ActionType,
+  instance: TableInstance<T>,
+) => {
+  const resetSelection = () => {
+    instance.rows.forEach((r) => r.toggleRowSelected(false));
+  };
+
+  const getGreaterId = (id: string) => {
+    const greaterId = id.split('.');
+    greaterId[greaterId.length - 1] = `${
+      parseInt(greaterId[greaterId.length - 1]) + 1
+    }`;
+    return greaterId.join('.');
+  };
+
+  const getParentId = (id: string) => {
+    const lastDotIndex = id.lastIndexOf('.');
+    if (lastDotIndex < 0) {
+      return null;
+    }
+    return id.substring(0, lastDotIndex);
+  };
+
+  // Returns true if endId is a parent of currentId
+  // E.g. currentId = 1.4.2.3, endId = 1.4  -> true
+  // E.g. currentId = 1.2.5, endId = 1.3    -> false
+  const isAncestorParent = (currentId: string, endId: string) => {
+    return endId.indexOf(currentId) === 0;
+  };
+
+  // Returns the next row as if it was just a non subrows list
+  const getNextRow = (currentRow: Row<T>, startId: string, endId: string) => {
+    let nextRow;
+    if (
+      (currentRow.subRows ?? []).length > 0 &&
+      isAncestorParent(currentRow.id, endId)
+    ) {
+      // child
+      nextRow = currentRow.subRows[0];
+    } else {
+      // next sibling
+      nextRow = instance.rowsById[getGreaterId(currentRow.id)];
+
+      // If the next sibling doesn't exist, go to the next child at a one upper depth
+      let parentId = getParentId(currentRow.id);
+      while (nextRow == null && parentId != null) {
+        nextRow = instance.rowsById[getGreaterId(parentId)];
+        parentId = getParentId(parentId);
+      }
+    }
+
+    return nextRow;
+  };
+
+  resetSelection();
+
+  // If no row selected before shift clicking, then assume first row is the last selectedRow (like Windows File Explorer)
+  let startId = state.lastSelectedRow ?? instance.rows[0].id;
+  let endId = action.id;
+
+  // Make sure startId is lesser than endId. Selection always goes from top to bottom
+  if (startId > endId) {
+    const temp = startId;
+    startId = endId;
+    endId = temp;
+  }
+
+  let currentRow = instance.rowsById[startId];
+
+  while (true) {
+    if (currentRow.id === endId) {
+      currentRow.toggleRowSelected(true);
+      break;
+    }
+
+    if (!isAncestorParent(currentRow.id, endId)) {
+      currentRow.toggleRowSelected(true);
+    }
+
+    currentRow = getNextRow(currentRow, startId, endId);
+  }
```

Comment:
Not sure if there is a way, but can you think of a more cleaner code for this? Or is this good enough?

Reply:
My thought process would be get all selected rows (from startId to endId) in array and then iterate through them to toggleRowSelected. Not sure if cleaner or better.

Reply:
Thanks, that's similar to as it is now. Currently, we start from `startId` and keep going to the next row/subrow until we reach the `endId`. At each node that we iterate through, we `toggleRowSelected(true)` (if applicable).